### PR TITLE
EN-13565: txSigner refactor

### PR DESCRIPTION
**Pull request by dragos-rebegea**

### CODE REVIEW

File: blockchain/txSigner.go [hunk 1]
```diff
@@ -1,41 +1,61 @@
...
+	suite         = ed25519.NewEd25519()
+	hasher        = keccak.NewKeccak()
+	singleSigner  = &singlesig.Ed25519Signer{}
+	messagePrefix = []byte("\x17Elrond Signed Message:\n")
+)
 
-// txSigner contains the primitives used to correctly sign a transaction
-type txSigner struct {
+// xSigner contains the primitives used to correctly sign a transaction
+type xSigner struct {
```
Comment (iulianpascalau) [inline blockchain/txSigner.go:19]: could have been simple `signer`

File: blockchain/txSigner.go [hunk 2]
```diff
@@ -1,41 +1,61 @@
 package blockchain
 
 import (
+	"github.com/ElrondNetwork/elrond-go-core/hashing/keccak"
```
Comment (iulianpascalau) [inline blockchain/txSigner.go:4]: do not forget to properly rename the file also

File: blockchain/txSigner.go [hunk 3]
```diff
@@ -1,41 +1,61 @@
...
+	return &xSigner{
 		keyGen: signing.NewKeyGenerator(suite),
 	}
 }
 
-// SignMessage will generate the signature providing the private key bytes and the message bytes (usually the serialized
-// form of the transaction)
-func (ts *txSigner) SignMessage(msg []byte, skBytes []byte) ([]byte, error) {
-	sk, err := ts.keyGen.PrivateKeyFromByteArray(skBytes)
+// SignMessage will generate the signature providing the private key bytes and the message bytes
```
Comment (iulianpascalau) [inline blockchain/txSigner.go:30]: L31-L54 can be refactored as follows:
```GO
// SignMessage will generate the signature providing the private key bytes and the message bytes
// prepending the standard messagePrefix
func (xs *xSigner) SignMessage(msg []byte, skBytes []byte) ([]byte, error) {
	serializedMessage, err := xs.serializeForSigning(msg)
	if err != nil {
		return nil, err
	}

	return xs.SignByteSlice(serializedMessage, skBytes)
}

// SignByteSlice will generate the signature providing the private key bytes and the byte slice to be signed
func (xs *xSigner) SignByteSlice(buff []byte, skBytes []byte) ([]byte, error) {
	sk, err := xs.keyGen.PrivateKeyFromByteArray(skBytes)
	if err != nil {
		return nil, err
	}

	return singleSigner.Sign(sk, buff)
}
```
you will improve code reuse and rename the somehow misleading `SignTransaction` function which is not related to a transaction at all.

File: builders/interface.go [hunk 4]
```diff
@@ -40,9 +40,10 @@ type VMQueryBuilder interface {
 	IsInterfaceNil() bool
 }
 
-// TxSigner defines the method used by a struct used to create valid signatures
-type TxSigner interface {
+// XSigner defines the method used by a struct used to create valid signatures
+type XSigner interface {
```
Comment (iulianpascalau) [inline builders/interface.go:44]: Signer instead of XSigner?

File: blockchain/txSigner.go [hunk 5]
```diff
@@ -1,41 +1,61 @@
 package blockchain
```
Comment (ccorcoveanu) [inline blockchain/txSigner.go:1]: maybe also rename the file

File: blockchain/txSigner.go [hunk 6]
```diff
@@ -45,7 +65,13 @@ func (ts *txSigner) GeneratePkBytes(skBytes []byte) ([]byte, error) {
 	return pk.ToByteArray()
 }
 
+func (xs *xSigner) serializeForSigning(msg []byte) ([]byte, error) {
+	msg = append(messagePrefix, msg...)
```
Comment (ccorcoveanu) [inline blockchain/txSigner.go:69]: I think you forgot about this: https://github.com/ElrondNetwork/elrond-sdk-erdjs/blob/main/src/signableMessage.ts#L46

Then - can we have a test that a signed message from js validates here? (plus verification as discussed in the call)

File: blockchain/txSigner.go (hunk 5 above)
Comment (dragos-rebegea) [reply blockchain/txSigner.go]: done

File: blockchain/txSigner.go (hunk 6 above)
Comment (dragos-rebegea) [reply blockchain/txSigner.go]: done

File: builders/interface.go (hunk 4 above)
Comment (dragos-rebegea) [reply builders/interface.go]: done

File: blockchain/txSigner.go (hunk 3 above)
Comment (dragos-rebegea) [reply blockchain/txSigner.go]: refactored

File: blockchain/txSigner.go (hunk 2 above)
Comment (dragos-rebegea) [reply blockchain/txSigner.go]: done

File: blockchain/txSigner.go (hunk 1 above)
Comment (dragos-rebegea) [reply blockchain/txSigner.go]: done

File: authentication/native_test.go [hunk 7]
```diff
@@ -5,26 +5,23 @@ import (
...
-	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
 func TestNativeAuthClient_NewNativeAuthClient(t *testing.T) {
 	t.Parallel()
 
-	expectedErr := errors.New("expected error")
-	t.Run("nil txsigner should error", func(t *testing.T) {
+	t.Run("nil XSigner should error", func(t *testing.T) {
```
Comment (iulianpascalau) [inline authentication/native_test.go:20]: nil Signer?

File: authentication/native_test.go [hunk 8]
```diff
@@ -126,12 +87,12 @@ func TestNativeAuthClient_GetAccessToken(t *testing.T) {
 		require.Equal(t, "", token)
 		require.Equal(t, expectedErr, err)
 	})
-	t.Run("txSigner errors when sign message", func(t *testing.T) {
+	t.Run("XSigner errors when sign message", func(t *testing.T) {
```
Comment (iulianpascalau) [inline authentication/native_test.go:90]: Signer errors?

File: blockchain/cryptoProvider/cryptoComponentsHolder.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+// GetPrivateKey returns the held privateKey
+func (holder *cryptoComponentsHolder) GetPrivateKey() crypto.PrivateKey {
+	return holder.privateKey
+}
+
+// GetBech32 returns the held bech32 address
+func (holder *cryptoComponentsHolder) GetBech32() string {
+	return holder.addressHandler.AddressAsBech32String()
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/cryptoComponentsHolder.go:55]: this could have been cached in a string that was created on the constructor

File: blockchain/cryptoProvider/interface.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+import (
+	crypto "github.com/ElrondNetwork/elrond-go-crypto"
+	"github.com/ElrondNetwork/elrond-sdk-erdgo/core"
+)
+
+// CryptoComponentsHolder is able to holder and provide all the crypto components
+type CryptoComponentsHolder interface {
+	GetPublicKey() crypto.PublicKey
+	GetPrivateKey() crypto.PrivateKey
+	GetBech32() string
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/interface.go:12]: GetBech32Address?

Also, this interface should be moved where used, not implemented :D

File: blockchain/cryptoProvider/signer.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+	hasher        = keccak.NewKeccak()
+	singleSigner  = &singlesig.Ed25519Signer{}
+	messagePrefix = []byte("\x17Elrond Signed Message:\n")
+)
+
+// signer contains the primitives used to correctly sign a transaction
+type signer struct{}
+
+// NewXSigner will create a new instance of signer
+func NewXSigner() *signer {
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/signer.go:25]: NewSigner?

File: blockchain/cryptoProvider/signer.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+	serializedMessage, err := xs.serializeForSigning(msg)
+	if err != nil {
+		return nil, err
+	}
+
+	return singleSigner.Sign(privateKey, serializedMessage)
+}
+
+// VerifyMessage will verify the signature providing the public key bytes and the message bytes
+func (xs *signer) VerifyMessage(msg []byte, publicKey crypto.PublicKey, sig []byte) error {
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/signer.go:41]: Does adding the following method `VerifyByteSlice` will improve the reusability of this component?

File: blockchain/cryptoProvider/signer.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+var (
+	signerLog     = logger.GetOrCreate("elrond-sdk-erdgo/signer")
+	hasher        = keccak.NewKeccak()
+	singleSigner  = &singlesig.Ed25519Signer{}
+	messagePrefix = []byte("\x17Elrond Signed Message:\n")
+)
+
+// signer contains the primitives used to correctly sign a transaction
+type signer struct{}
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/signer.go:22]: unit tests?

File: builders/txBuilder.go [hunk 14]
```diff
@@ -2,38 +2,34 @@ package builders
...
 var (
-	log                    = logger.GetOrCreate("elrond-sdk-erdgo/builders")
-	txHasher               = keccak.NewKeccak()
 	blake2bHasher          = blake2b.NewBlake2b()
 	nodeInternalMarshaller = &marshal.GogoProtoMarshalizer{}
 )
 
 type txBuilder struct {
-	txSigner TxSigner
+	xSigner Signer
```
Comment (iulianpascalau) [inline builders/txBuilder.go:22]: signer instead of xSigner?

File: builders/txBuilder.go [hunk 15]
```diff
@@ -57,28 +53,16 @@ func (builder *txBuilder) createTransaction(arg data.ArgCreateTransaction) *data
 // ApplySignatureAndGenerateTx will apply the corresponding sender and compute the signature field and
 // generate the transaction instance
 func (builder *txBuilder) ApplySignatureAndGenerateTx(
```
Comment (iulianpascalau) [inline builders/txBuilder.go:55]: :+1:

File: builders/txBuilder.go [hunk 16]
```diff
@@ -146,11 +130,11 @@ func transactionToNodeTransaction(tx *data.Transaction) (*transaction.Transactio
 	}, nil
 }
 
-func (builder *txBuilder) createUnsignedMessage(arg data.ArgCreateTransaction) ([]byte, error) {
+func (builder *txBuilder) createUnsignedTx(arg data.ArgCreateTransaction) (*data.Transaction, error) {
 	arg.Signature = ""
 	tx := builder.createTransaction(arg)
 
-	return json.Marshal(tx)
+	return tx, nil
```
Comment (iulianpascalau) [inline builders/txBuilder.go:137]: this function never returns a non-nil error. Consider removing the error output parameter.

File: testsCommon/xSignerStub.go [hunk 17]
```diff
@@ -0,0 +1,55 @@
+package testsCommon
```
Comment (iulianpascalau) [inline testsCommon/xSignerStub.go:1]: rename the file to signerStub.go?

**Review by iulianpascalau — COMMENTED** [review]
Very good refactoring :+1: 
Minor changes required.

File: blockchain/cryptoProvider/signer.go (hunk 12 above)
Comment (ccorcoveanu) [reply blockchain/cryptoProvider/signer.go:41]: Yes, I would say that if we expose `SignByteSlice` we should also expose `VerifyByteSlice`

File: blockchain/cryptoProvider/cryptoComponentsHolder.go (hunk 9 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/cryptoComponentsHolder.go:55]: done

File: testsCommon/xSignerStub.go (hunk 17 above)
Comment (dragos-rebegea) [reply testsCommon/xSignerStub.go:1]: done

File: builders/txBuilder.go (hunk 16 above)
Comment (dragos-rebegea) [reply builders/txBuilder.go:136]: done

File: builders/txBuilder.go (hunk 14 above)
Comment (dragos-rebegea) [reply builders/txBuilder.go:21]: done

File: authentication/native_test.go (hunk 7 above)
Comment (dragos-rebegea) [reply authentication/native_test.go:20]: done

File: authentication/native_test.go (hunk 8 above)
Comment (dragos-rebegea) [reply authentication/native_test.go:90]: done

File: blockchain/cryptoProvider/interface.go (hunk 10 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/interface.go]: done

File: blockchain/cryptoProvider/signer.go (hunk 11 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/signer.go:25]: Done

File: blockchain/cryptoProvider/signer.go (hunk 13 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/signer.go:22]: done

File: blockchain/cryptoProvider/signer.go (hunk 12 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/signer.go:45]: done

File: blockchain/cryptoProvider/cryptoComponentsHolder.go [hunk 18]
```diff
@@ -2,41 +2,33 @@ package cryptoProvider
...
 	if err != nil {
 		return nil, err
 	}
 	publicKey := privateKey.GeneratePublic()
-
-	publicKeyBytes, err := publicKey.ToByteArray()
-	if err != nil {
-		return nil, err
-	}
+	publicKeyBytes, _ := publicKey.ToByteArray()
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/cryptoComponentsHolder.go:23]: I know this can not return an error, as the current implementations look. However, masking this can lead to some unexpected behavior. Suggestion to revert the changes here.

File: blockchain/cryptoProvider/signer.go [hunk 19]
```diff
@@ -19,27 +19,31 @@ var (
...
+// NewSigner will create a new instance of signer
+func NewSigner() *signer {
+	return &signer{
+		serializeForSigningHandle: serializeForSigning,
+	}
 }
 
 // SignMessage will generate the signature providing the private key bytes and the message bytes
 // prepending the standard messagePrefix
 func (xs *signer) SignMessage(msg []byte, privateKey crypto.PrivateKey) ([]byte, error) {
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/signer.go:35]: `s *signer` ?

File: blockchain/cryptoProvider/signer_test.go [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+		sig, err := signer.SignMessage([]byte("msg"), &testsCommon.PrivateKeyStub{})
+		require.Nil(t, sig)
+		require.Equal(t, expectedError, err)
+	})
+	t.Run("should work", func(t *testing.T) {
+		t.Parallel()
+
+		signer := NewSigner()
+		sk, _ := hex.DecodeString("45f72e8b6e8d10086bacd2fc8fa1340f82a3f5d4ef31953b463ea03c606533a6")
+		privateKey, err := keyGen.PrivateKeyFromByteArray(sk)
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/signer_test.go:41]: from linter:
```
Check failure on line 41 in blockchain/cryptoProvider/signer_test.go

GitHub Actions
/ golangci linter

ineffectual assignment to err (ineffassign)
```

File: blockchain/cryptoProvider/signer_test.go [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+func TestSigner_VerifyByteSlice(t *testing.T) {
+	t.Parallel()
+
+	t.Run("should work", func(t *testing.T) {
+		t.Parallel()
+
+		msg := []byte("msg")
+		signer := NewSigner()
+		sk, _ := hex.DecodeString("45f72e8b6e8d10086bacd2fc8fa1340f82a3f5d4ef31953b463ea03c606533a6")
+		privateKey, err := keyGen.PrivateKeyFromByteArray(sk)
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/signer_test.go:72]: from linter:
```GO
 Check failure on line 72 in blockchain/cryptoProvider/signer_test.go

GitHub Actions
/ golangci linter

ineffectual assignment to err (ineffassign)
```

File: blockchain/cryptoProvider/signer.go [hunk 22]
```diff
@@ -19,27 +19,31 @@ var (
...
+	serializeForSigningHandle func(msg []byte) ([]byte, error)
+}
 
-// NewXSigner will create a new instance of signer
-func NewXSigner() *signer {
-	return &signer{}
+// NewSigner will create a new instance of signer
+func NewSigner() *signer {
+	return &signer{
+		serializeForSigningHandle: serializeForSigning,
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/signer.go:29]: why did we need this handler? It is so hard to test the serializeForSigning function in the test? It does only some converts, byte slice appends and a hash.Compute operation

File: blockchain/cryptoProvider/signer_test.go [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+		signer.serializeForSigningHandle = func(msg []byte) ([]byte, error) {
+			return nil, expectedError
+		}
+		require.False(t, check.IfNil(signer))
+		tx := &data.Transaction{Version: 2, Options: 1}
+		sig, err := signer.SignTransaction(tx, privateKey)
+		require.NotNil(t, sig)
+		require.Nil(t, err)
+	})
+}
```
Comment (ccorcoveanu) [inline blockchain/cryptoProvider/signer_test.go:130]: Some extra tests for SignMessage maybe? Also, I think a test to cross check a signed message from the JS sdk also verifies here would be nice to ensure compatibility.

File: blockchain/cryptoProvider/signer.go (hunk 19 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/signer.go:35]: done

File: blockchain/cryptoProvider/signer_test.go (hunk 20 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/signer_test.go:41]: done

File: blockchain/cryptoProvider/signer_test.go (hunk 21 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/signer_test.go:72]: done

File: blockchain/cryptoProvider/signer.go (hunk 22 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/signer.go:29]: indeed, it was not necessary. removed

File: blockchain/cryptoProvider/signer_test.go (hunk 23 above)
Comment (dragos-rebegea) [reply blockchain/cryptoProvider/signer_test.go:122]: added

File: blockchain/txSigner.go (hunk 6 above)
Comment (dragos-rebegea) [reply blockchain/txSigner.go]: added also the test for that

File: blockchain/cryptoProvider/signer_test.go [hunk 24]
```diff
@@ -118,13 +91,32 @@ func TestSigner_SignTransaction(t *testing.T) {
...
 		require.Nil(t, err)
 	})
+	// {
+	// 	 "address":"erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th", //alice
+	//	 "message":"0x6d657373616765",
+	//	 "signature":"0x546c6b6d6487852f54571ab2da81b48ff8f09bef71ba07b116fcf7203538cd64ea5f9bffcc13a0279a75ca3b1b0a1e478d23e1771d381011f8135e4372a9dd00",
+	//	 "version":1,
+	//	 "signer":"ErdJS"
+	// }
+	t.Run("should work with signature generated using erdjs", func(t *testing.T) {
```
Comment (iulianpascalau) [inline blockchain/cryptoProvider/signer_test.go:107]: :+1:

**Review by iulianpascalau — APPROVED** [review]

**Review by ccorcoveanu — APPROVED** [review]
